ONSdigital/rm-action-service
# Fix duplicate action creation & improve performance

# Motivation and Context
There is a production bug where actions are being created and sent twice, because there are 2 instances of the Action service in production, but the locking mechanism and transaction boundaries are not correctly preventing each instance from performing the same work.

# What has changed
The transaction which creates the actions must now be committed - in its own new transaction - before the Redis lock is released. This stops the problem where two instances can create the same actions, solving the production bug.

Unnecessary round-tripping to the database to get the ActionType entity has been replaced with code which caches these entities in memory for the duration of the action creation process.

Instead of flushing to the DB after every single action is created - using `saveAndFlush` - now each new action is added to Hibernate's cache using the `save` repo function. Once all the actions have been created, they are flushed to the DB using the `flush` repo function. This is vastly more performance efficient (a few minutes versus a few hours).

Testing on my local machine, the new code created 3,450 actions in 76 seconds. The old code took at least 10 minutes, so this is a substantial performance gain, and capable of creating 350,000 actions in circa 2 hours (versus 14 hours).

# How to test?
Run the acceptance tests. The actions should be created and cause Action Exporter and Notify Gateway to be invoked to send emails and letters, as normal.

To test the performance improvement, load a large sample into a collection exercise, wait for all the cases to be created in the Action database, then change the trigger time on an appropriate action so that the Action service starts creating the actions.

# Links
Trello: https://trello.com/c/IETXD17d/551-bug-action-service-creates-duplicate-actions-as-part-of-the-actionplanjobexecutor
